File: kdm6_torch/kdm6/cloud_dsd.py
"""
KDM6 cloud DSD diagnostics — B1-B4의 placeholder 의존성 해소 모듈.

원본:
  - module_mp_kdm6.F: 1670-1673 (avedia_c, avedia_r, sigma_c)
  - module_mp_kdm6.F: 1703-1705 (lencon, lenconcr)
  - module_mp_kdm6.F: 842-847   (qcr - sea/land 분기)
  - module_mp_kdm6.F: 3104-3105 (qc0, qc1)
  - module_mp_kdm6.F: 3173      (pidnc = cmc * rgmma(1+dmc/(muc+1)))
  - module_mp_kdm6.F: 770       (lamdac = (pidnc·nc / (qc·den))^(1/dmc))

본 모듈이 산출하는 5개 양은 warm.py의 B1-B4 함수들이 입력으로 받는 placeholder.
Step B 검증 후 B1-B4를 *coordinator*에서 호출할 때 본 진단을 먼저 수행.
"""
from __future__ import annotations

from math import exp, lgamma, pi as _pi
from typing import NamedTuple

import torch

from . import constants as c
from . import fconst as _fc


def _rgmma(x: float) -> float:
    """Fortran `rgmma(x) = exp(GAMMLN(x)) = Γ(x)` 직역. GAMMLN은 ln(Γ)이므로
    rgmma는 *Γ(x)* (1/Γ 아님). 이전 구현은 `exp(-lgamma)` = 1/Γ로 잘못 짜여
    review6 audit에서 발견."""
    # Fortran rgmma = f32 expf(f32 gammln) — differs from exp(lgamma) at non-integer args (step-67 class)
    return _fc.rgmma_f(x)


# ─── Params ──────────────────────────────────────────────────────────────────


class CloudDsdParams(NamedTuple):
    pidnc: float       # cmc · rgmma(1 + dmc/(muc+1))
    dmc: float
    muc: float
    lamdacmax: float
    lamdacmin: float
    g3pmc: float       # rgmma(1 + 3/(muc+1))
    g6pmc: float       # rgmma(1 + 6/(muc+1))
    g4pmr_over_g1pmr: float  # rgmma(4+mur)/rgmma(1+mur)
    qc0: float         # continental critical: 4/3·π·denr·r0³·xncr0/den0
    qc1: float         # maritime    critical: 4/3·π·denr·r0³·xncr1/den0


def default_cloud_dsd_params(*, den0: float | None = None) -> CloudDsdParams:
    """Fortran kdm6init의 cloud-side 파생 상수.

    den0 default = c.DEN0 (1.28 kg/m³, 표준대기).
    """
    if den0 is None:
        den0 = c.DEN0
    cmc = _pi * c.DENR / 6.0
    pidnc = _fc.PIDNC   # f32-stepwise (kdm6init F:3205; see fconst.py)
    g3pmc = _rgmma(1.0 + 3.0 / (c.MUC + 1.0))
    g6pmc = _rgmma(1.0 + 6.0 / (c.MUC + 1.0))
    g1pmr = _rgmma(1.0 + c.MUR)
    g4pmr = _rgmma(4.0 + c.MUR)

    # STEP-91 latent class mirror: kdm6init F:3135-3136 REAL(4) l2r — *xncr BEFORE /den0.
    _qc_pre = _fc._f32(_fc._f32(_fc._f32(_fc._f32(4.0 / 3.0) * _fc.PI) * _fc._f32(c.DENR)) * _fc.powf(c.R0, 3.0))
    qc0 = _fc._f32(_fc._f32(_qc_pre * _fc._f32(c.XNCR0)) / _fc._f32(den0))
    qc1 = _fc._f32(_fc._f32(_qc_pre * _fc._f32(c.XNCR1)) / _fc._f32(den0))

    return CloudDsdParams(
        pidnc=pidnc,
        dmc=c.DMC,
        muc=c.MUC,
        lamdacmax=c.LAMDACMAX,
        lamdacmin=c.LAMDACMIN,
        g3pmc=g3pmc,
        g6pmc=g6pmc,
        g4pmr_over_g1pmr=g4pmr / g1pmr,
        qc0=qc0,
        qc1=qc1,
    )


# ─── Cloud slope (rslopec) ────────────────────────────────────────────────────


def diag_cloud_slope_torch(
    qc: torch.Tensor,
    nc: torch.Tensor,
    den: torch.Tensor,
    *,
    params: CloudDsdParams,
    ncmin_tensor: "torch.Tensor | None" = None,
) -> torch.Tensor:
    """rslopec = 1/lamdac, clamp to [1/lamdacmax, 1/lamdacmin], with the Fortran
    inactive-cloud gate (nc<=ncmin → rslopecmax = 1/lamdacmax).

    lamdac = (pidnc·nc / max(qc·den, 1e-30))^(1/dmc)
    """
    DOMAIN_FLOOR = 1.0e-30
    ratio = params.pidnc * nc / torch.clamp(qc * den, min=DOMAIN_FLOOR)
    lamdac = torch.exp(torch.log(torch.clamp(ratio, min=DOMAIN_FLOOR)) * _fc._f32(1.0 / _fc._f32(params.dmc)))  # Fortran *(1./dmc) — step-65 class
    # STEP-75 SEED (D-B) mirror: Fortran's ACTIVE rslopec is UNCLAMPED (stmt fn
    # lamdac F:802 has no bounds; the lamdacmax SNAP rewrites only nci/n0c). The
    # NaN guard is preserved structurally: the 1e-30 domain floors bound lamdac > 0
    # (no Inf), and the explosive qc/nc~0 cells are exactly the INACTIVE set
    # overwritten below with 1/lamdacmax (Fortran F:1454 same branch).
    rslopec_active = 1.0 / lamdac
    # Fortran F:1603-1608 inactive-cloud branch: (qc<=qmin .or. nc<=ncmin) →
    # rslopec = rslopecmax = 1/lamdacmax. Per-cell ncmin (xland) via ncmin_tensor.
    nc_floor = ncmin_tensor if ncmin_tensor is not None else c.NCMIN
    inactive = (qc <= c.EPS) | (nc <= nc_floor)
    rslopec = torch.where(inactive,
                          torch.full_like(rslopec_active, 1.0 / params.lamdacmax),
                          rslopec_active)
    return rslopec


def diag_species_slope_torch(
    q: torch.Tensor,
    n: torch.Tensor,
    den: torch.Tensor,
    pidn: float,
    dm: float,
    lamdamax: float,
    lamdamin: float,
) -> torch.Tensor:
    """rslope = 1/lamda clamped to [1/lamdamax, 1/lamdamin] (rain/ice species).

    lamda = (pidn·n / max(q·den, 1e-30))^(1/dm). 1:1 mirror of C++
    cloud_dsd::diag_species_slope_torch (the clamped DSD slope). Used by
    build_default_aux_torch for rslope_r/rslope_i. AD-safe (clamp/exp/log only).
    """
    DOMAIN_FLOOR = 1.0e-30
    ratio = pidn * n / torch.clamp(q * den, min=DOMAIN_FLOOR)
    # Fortran multiplies by the REAL(4) reciprocal `*(1./dmc)` — not /dm (step-65
    # lamda evaluation-form class; mirrors C++ cloud_dsd.cpp).
    inv_dm = _fc._f32(1.0 / _fc._f32(dm))
    lamda = torch.exp(torch.log(torch.clamp(ratio, min=DOMAIN_FLOOR)) * inv_dm)
    return torch.clamp(1.0 / lamda, min=1.0 / lamdamax, max=1.0 / lamdamin)


# ─── avedia (cloud, rain) ─────────────────────────────────────────────────────


def diag_avedia_cloud_torch(rslopec: torch.Tensor, *, params: CloudDsdParams) -> torch.Tensor:
    """Fortran 1670: avedia_c = rslopec * g3pmc^(1/3)."""
    return rslopec * (params.g3pmc ** (1.0 / 3.0))


def diag_avedia_rain_torch(rslope_r: torch.Tensor, *, params: CloudDsdParams) -> torch.Tensor:
    """Fortran 1671: avedia_r = rslope_r * (g4pmr/g1pmr)^.3333333 (truncated literal)."""
    return rslope_r * (params.g4pmr_over_g1pmr ** 0.3333333)  # 1:1 fix #4 (cloud avedia uses 1/3, see line 131)


# ─── sigma (cloud DSD width) ──────────────────────────────────────────────────


def diag_sigma_cloud_torch(
    rslopec: torch.Tensor,
    *,
    params: CloudDsdParams,
    strict_fortran: bool = False,
) -> torch.Tensor:
    """Fortran 1673: sigma_c = rslopec * (g6pmc - g3pmc²)^(1/6).

    review6 audit 후 `rgmma = Γ` 직역 적용 → muc=2에서:
        g3pmc = Γ(2) = 1.0, g6pmc = Γ(3) = 2.0, var_factor = 1.0 (positive).
    이전 코드의 `strict_fortran` mode와 EPS clamp는 잘못된 reciprocal-gamma 가정의
    잔재. 이제 Fortran 직역 자체가 안정. clamp는 numerical safety guard로만 유지.
    """
    raw = params.g6pmc - params.g3pmc * params.g3pmc
    if strict_fortran:
        var_factor = raw
    else:
        var_factor = max(raw, 1.0e-30)
    return rslopec * (var_factor ** (1.0 / 6.0))


# ─── lencon, lenconcr ─────────────────────────────────────────────────────────


def diag_lencon_torch(
    qc: torch.Tensor,
    den: torch.Tensor,
    avedia_c: torch.Tensor,
    sigma_c: torch.Tensor,
    *,
    qcrmin: float = c.QCRMIN,
) -> tuple[torch.Tensor, torch.Tensor]:
    """Fortran 1703-1705:
        lencon = 2.7e-2 · den · qc · (1e20/16 · avedia_c · sigma_c³ - 0.4)
        lenconcr = max(1.2 · lencon, qcrmin)

    autoconv → accretion 전환의 임계 값. 음수가 될 수 있어서 후행 max로 가드.
    """
    factor = 1.0e20 / 16.0 * avedia_c * (sigma_c ** 3) - 0.4
    lencon = 2.7e-2 * den * qc * factor
    lenconcr = torch.clamp(1.2 * lencon, min=qcrmin)
    return lencon, lenconcr


# ─── qcr (sea/land 분기) ──────────────────────────────────────────────────────


def diag_qcr_torch(
    sea_mask: torch.Tensor,
    *,
    params: CloudDsdParams,
    ref: torch.Tensor | None = None,
) -> torch.Tensor:
    """Fortran module_mp_kdm6.F:842-847 — sea(slmsk==2) → qc0, land → qc1.

    Physical reasoning: qc0 = qc_base · XNCR0 (XNCR0=5e7, low CCN concentration)
    and qc1 = qc_base · XNCR1 (XNCR1=5e8, high CCN). Higher CCN → smaller
    cloud droplets → harder autoconversion → HIGHER qcr threshold. Ocean air
    is clean (LOW CCN) so it gets the LOWER threshold qc0; land air is dusty
    (HIGH CCN) so it gets the HIGHER threshold qc1. The Param-field names
    `qc0/continental`, `qc1/maritime` in CloudDsdParams are legacy labels
    pinned to the scalar values, not the regime mapping; the regime wiring
    is here and mirrors the operational Fortran assignment.

    Parameters
    ----------
    sea_mask : (B, K) boolean (True = sea)
    ref      : float tensor whose dtype/device the output should match.
               If None, defaults to float64/cpu (legacy behavior).

    Returns
    -------
    qcr : (B, K) tensor with qc0 (sea) or qc1 (land).
    """
    if ref is None:
        dtype = torch.float64
        device = sea_mask.device
    else:
        dtype = ref.dtype
        device = ref.device
    qc0_t = torch.tensor(params.qc0, dtype=dtype, device=device)
    qc1_t = torch.tensor(params.qc1, dtype=dtype, device=device)
    qcr = torch.where(sea_mask, qc0_t, qc1_t)
    return qcr


__all__ = [
    "CloudDsdParams",
    "default_cloud_dsd_params",
    "diag_cloud_slope_torch",
    "diag_avedia_cloud_torch",
    "diag_avedia_rain_torch",
    "diag_sigma_cloud_torch",
    "diag_lencon_torch",
    "diag_qcr_torch",
]

File: kdm6_libtorch/src/cloud_dsd.cpp
// KDM6 cloud DSD diagnostics — C++ port of kdm6_torch/kdm6/cloud_dsd.py.
// review6 audit fix: rgmma = Γ(x) (= exp(lgamma(x))).
//
#include "kdm6/cloud_dsd.h"
#include "kdm6/fconst.h"
#include "kdm6/ops.h"

#include <cmath>

namespace kdm6 {
namespace cloud_dsd {

namespace {

constexpr double PI = 3.14159265358979323846;

// Fortran rgmma(x) = exp(GAMMLN(x)) = Γ(x). review6 audit 부호 수정.
inline double rgmma_scalar(double x) {
    // Fortran rgmma(x)=EXP(GAMMLN(x)) in REAL(4): f32 expf of the f32-rounded
    // double Lanczos — differs from exp(lgamma(double)) for NON-INTEGER args
    // (e.g. Γ(4/3), Γ(7/3) in D2/D3 freezing — the step-67 qi/ni seed class).
    return static_cast<double>(fconst::rgmma_f(static_cast<float>(x)));
}

constexpr double DOMAIN_FLOOR = 1.0e-30;

}  // namespace

CloudDsdParams default_cloud_dsd_params(double den0) {
    const double cmc = PI * constants::DENR / 6.0;
    const double pidnc = fconst::get().pidnc;   // f32-stepwise (kdm6init F:3205)
    const double g3pmc = fconst::get().g3pmc;   // f32-stepwise
    const double g6pmc = fconst::get().g6pmc;   // f32-stepwise
    const double g1pmr = fconst::get().g1pmr;   // f32-stepwise
    const double g4pmr = fconst::get().g4pmr;   // f32-stepwise

    // STEP-91 latent class: kdm6init F:3135-3136 builds qc0/qc1 in REAL(4)
    // left-to-right — 4./3.*pi*denr*r0**3*xncr/den0 (MULTIPLY by xncr BEFORE
    // dividing by den0; r0**3 is f32 powf). Double chain + reordered ops can
    // differ; these gate praut (qcr threshold).
    const float pi_qc = static_cast<float>(fconst::get().pi);
    const float r0_f = static_cast<float>(constants::R0);
    const float r03_f = std::pow(r0_f, 3.0f);
    const float qc_pre = (((4.0f / 3.0f) * pi_qc) * static_cast<float>(constants::DENR)) * r03_f;
    const double qc0 = static_cast<double>(
        (qc_pre * static_cast<float>(constants::XNCR0)) / static_cast<float>(den0));
    const double qc1 = static_cast<double>(
        (qc_pre * static_cast<float>(constants::XNCR1)) / static_cast<float>(den0));

    return CloudDsdParams{
        /*pidnc=*/pidnc,
        /*dmc=*/constants::DMC,
        /*muc=*/constants::MUC,
        /*lamdacmax=*/constants::LAMDACMAX,
        /*lamdacmin=*/constants::LAMDACMIN,
        /*g3pmc=*/g3pmc,
        /*g6pmc=*/g6pmc,
        /*g4pmr_over_g1pmr=*/g4pmr / g1pmr,
        /*qc0=*/qc0,
        /*qc1=*/qc1,
    };
}

torch::Tensor diag_species_slope_torch(
    const torch::Tensor& q,
    const torch::Tensor& n,
    const torch::Tensor& den,
    double pidn,
    double dm,
    double lamdamax,
    double lamdamin
) {
    auto qden_safe = torch::clamp(q * den, /*min=*/DOMAIN_FLOOR);
    auto ratio = pidn * n / qden_safe;
    // Fortran lamdac(x,y,z)=exp(log((pidnc*z)/(x*y))*(1./dmc)); route exp/log through libm
    // so the float32 forward bit-matches gfortran's libm.
    // Fortran multiplies by the REAL(4) reciprocal `*(1./dmc)` — NOT a division by
    // dm. log(ratio)~36 amplifies the f32(1/3)-vs-true-1/3 difference through exp
    // into ~13 ULP of lamda (the step-65 rslopec seed). Hold the f32 reciprocal in
    // a double so the scalar demotion reproduces gfortran exactly.
    const double inv_dm = static_cast<double>(1.0f / static_cast<float>(dm));
    auto lamda = ops::libm_exp(ops::libm_log(torch::clamp(ratio, /*min=*/DOMAIN_FLOOR)) * inv_dm);
    return torch::clamp(
        1.0 / lamda,
        /*min=*/1.0 / lamdamax,
        /*max=*/1.0 / lamdamin
    );
}

torch::Tensor diag_cloud_slope_torch(
    const torch::Tensor& qc,
    const torch::Tensor& nc,
    const torch::Tensor& den,
    const CloudDsdParams& p,
    const c10::optional<torch::Tensor>& ncmin_tensor
) {
    // STEP-75 SEED (D-B): the Fortran ACTIVE cloud slope is UNCLAMPED — stmt fn
    // lamdac (F:802) has NO bounds, and F:1454-1466 stores rslopec=1/lamdac as-is;
    // the lamdacmax SNAP (F:1485-1497) rewrites only nci/n0c, never the rslopec*
    // arrays. The previous clamp [1/lamdacmax,1/lamdacmin] (parity tracker #6) was
    // a NaN guard for the mask-multiply port; it diverged D2/D3 (rslopec^4..^9) in
    // the lamdac>lamdacmax snap regime. The guard is preserved structurally: the
    // 1e-30 domain floors bound lamdc >= exp(log(1e-30)/3) ~ 1e-10 > 0 (no Inf),
    // and the explosive qc/nc~0 cells are exactly the INACTIVE set overwritten
    // below with 1/lamdacmax (Fortran F:1454 takes the same branch).
    auto qden = torch::clamp(qc * den, /*min=*/DOMAIN_FLOOR);
    auto ratio = torch::clamp(p.pidnc * nc / qden, /*min=*/DOMAIN_FLOOR);
    const double inv_dmc = static_cast<double>(1.0f / static_cast<float>(p.dmc));
    auto lamdc = ops::libm_exp(ops::libm_log(ratio) * inv_dmc);
    auto rslopec_active = 1.0 / lamdc;                  // F:1461 ACTIVE, UNCLAMPED
    // Fortran F:1603-1608 inactive-cloud branch: (qc<=qmin .or. nc<=ncmin) →
    // rslopec = rslopecmax = 1/lamdacmax. Per-cell ncmin via ncmin_tensor.
    auto inactive = (qc <= constants::EPS) |
        (ncmin_tensor.has_value() ? (nc <= ncmin_tensor.value())
                                  : (nc <= constants::NCMIN));
    return torch::where(inactive,
                        torch::full_like(rslopec_active, 1.0 / p.lamdacmax),
                        rslopec_active);
}

torch::Tensor diag_avedia_cloud_torch(
    const torch::Tensor& rslopec,
    const CloudDsdParams& p
) {
    return rslopec * std::pow(p.g3pmc, 1.0 / 3.0);
}

torch::Tensor diag_avedia_rain_torch(
    const torch::Tensor& rslope_r,
    const CloudDsdParams& p
) {
    // Fortran F:1671 rain avedia uses the truncated literal `.3333333` (NOT 1./3.); cloud
    // avedia F:1670 DOES use 1./3. (see line 83). 1:1 parity fix #4.
    return rslope_r * std::pow(p.g4pmr_over_g1pmr, 0.3333333);
}

torch::Tensor diag_sigma_cloud_torch(
    const torch::Tensor& rslopec,
    const CloudDsdParams& p
) {
    const double raw = p.g6pmc - p.g3pmc * p.g3pmc;
    const double var_factor = std::max(raw, 1.0e-30);
    return rslopec * std::pow(var_factor, 1.0 / 6.0);
}

LenconOutputs diag_lencon_torch(
    const torch::Tensor& qc,
    const torch::Tensor& den,
    const torch::Tensor& avedia_c,
    const torch::Tensor& sigma_c,
    double qcrmin
) {
    // Fortran F:1703-1704 lencon = 2.7e-2*den*qci*(1.e20/16.*avedia*(sigma**3)-0.4):
    // (1e20/16)*avedia rounds, *(sigma**3) rounds, then the -0.4 add rounds —
    // strict source order under -ffp-contract=off (fma_acc is two-rounding).
    // sigma**3: gfortran expands the f32 integer power as (sigma*sigma)*sigma
    // (repeated multiplication, objdump-verified — NOT libm powf, which
    // differs on ~26% of f32 samples; IEEE sweep finding).
    auto coef = 1.0e20 / 16.0 * avedia_c;
    auto sigma3 = sigma_c * sigma_c * sigma_c;
    auto factor = ops::fma_acc(torch::full_like(coef, -0.4), coef, sigma3);
    auto lencon = 2.7e-2 * den * qc * factor;
    auto lenconcr = torch::clamp(1.2 * lencon, /*min=*/qcrmin);
    return LenconOutputs{/*lencon=*/lencon, /*lenconcr=*/lenconcr};
}

torch::Tensor diag_qcr_torch(
    const torch::Tensor& sea_mask,
    const CloudDsdParams& p,
    const torch::Tensor& ref
) {
    // Mirrors operational Fortran module_mp_kdm6.F:842-847 and Python oracle
    // kdm6_torch/kdm6/cloud_dsd.py:diag_qcr_torch:
    //   sea (slmsk==2)  → qc0 (low CCN → low qcr threshold, clean marine air)
    //   land (else)     → qc1 (high CCN → high qcr threshold, dusty air)
    // The Param-field names `qc0/continental`, `qc1/maritime` in CloudDsdParams
    // are legacy labels pinned to scalar values, not the regime mapping; the
    // regime wiring is here.
    auto opts = ref.options();
    auto qc0_t = torch::tensor(p.qc0, opts);
    auto qc1_t = torch::tensor(p.qc1, opts);
    return torch::where(sea_mask, qc0_t, qc1_t);
}

}  // namespace cloud_dsd
}  // namespace kdm6
